src/analysis/souffle/check_predicate.dl

```diff
@@ -0,0 +1,82 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+#define SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+
+#include "authorization_logic.dl"
+#include "dataflow_graph.dl"
+#include "tags.dl"
+
+.type Bool <: number
+
+.type CheckPredicate =
+  CP_TagPresence { ap: AccessPath, principal: Principal, tag: Tag }
```

Comment:
Just FYI:
- the error messages when using ADTs can be a bit cryptic at times (you might have already run into this :)
- ADTs are not a usual Datalog feature, they're Souffle-specific. So if we pivot to rolling our own solver, the ADTs might be less portable to our own solver.

You shouldn't necessarily let these points scare you off from using ADTs, but I just thought I'd let you know.

Reply:
Yes, that's true. If we switched to another Datalog dialect with lists and not ADTs, we could definitely just represent the ADTs as lists with a little less compiler support. If we switched to one without even that, we might be forced to do something ugly by using IDs to point across relations. But I think it is possible with plain ol' Datalog, so I'm not too worried about it for now.